strawberryfields/io/xir_io.py

```diff
@@ -0,0 +1,294 @@
+                        if isinstance(p, Decimal):
+                            params.append(float(p))
+                        elif isinstance(p, Iterable):
+                            params.append(np.array(_listr(p)))
+                        else:
+                            params.append(p)
+                    params = sfpar.par_convert(params, prog)
+                    gate(*params) | regrefs  # pylint:disable=expression-not-assigned
+            else:
+                if callable(gate):
+                    gate() | regrefs  # pylint:disable=expression-not-assigned,pointless-statement
+                else:
+                    gate | regrefs  # pylint:disable=expression-not-assigned,pointless-statement
+
+    prog._target = xir_prog.options.get("target", None)  # pylint: disable=protected-access
+
+    if "shots" in xir_prog.options:
+        prog.run_options["shots"] = xir_prog.options["shots"]
+    if "cutoff_dim" in xir_prog.options:
+        prog.backend_options["cutoff_dim"] = xir_prog.options["cutoff_dim"]
+
+    return prog
+
+
+# pylint: disable=too-many-branches
+def from_xir_to_tdm(xir_prog):
+    prog = TDMProgram(xir_prog.options["N"], name=xir_prog.options.get("name", "xir"))
+
+    # extract the tdm gate arguments from the xir program constants
+    args = [val for key, val in xir_prog.constants.items() if key[0] == "p" and key[1:].isdigit()]
+
+    # convert arguments to float/complex if stored as Decimal/DecimalComplex objects
+    for i, params in enumerate(args):
+        for j, p in enumerate(params):
+            if isinstance(p, Decimal):
+                args[i][j] = float(p)
+            elif isinstance(p, xir.DecimalComplex):
+                args[i][j] = complex(p)
+
+    # append the quantum operations
+    with prog.context(*args) as (p, q):
+        for op in xir_prog.statements:
+            # check if operation name is in the list of
+            # defined StrawberryFields operations.
+            # This is used by checking against the ops.py __all__
+            # module attribute, which contains the names
+            # of all defined quantum operations
+            if op.name in ops.__all__:
+                # get the quantum operation from the sf.ops module
+                gate = getattr(ops, op.name)
+            else:
+                raise NameError(f"Quantum operation {op.name!r} not defined!")
+
+            # create the list of regrefs
+            regrefs = [q[i] for i in op.wires]
+
+            if op.params:
+                # convert symbolic expressions to symbolic expressions containing the corresponding
+                # MeasuredParameter and FreeParameter instances.
+                if isinstance(op.params, dict):
+                    vals = sfpar.par_convert(op.params.values(), prog)
+                    params = dict(zip(op.params.keys(), vals))
+                    for key, val in params.items():
+                        if val[0] == "p" and val[1:].isdigit():
```

Comment:
This check should be a function (it's not obvious and we use it more than once).